### ISSUE TITLE
Feature Request: Quantization Control Plane

### ISSUE THREAD

**Issue by ahb-sjsu** [ahb-sjsu/turboquant-pro#169]
# Feature Request: Quantization Control Plane

## Summary

TurboQuant-Pro should evolve from a library that provides strong compression methods into a **quantization and compression control plane**:

> **Given a consumer, workload, resource budget, and available codecs, determine what should be compressed, how it should be compressed, whether the resulting representation preserves what the consumer actually needs, and when the system should fall back or re-plan.**

The goal is not for TurboQuant-Pro to implement the winning primitive codec in every regime.

The goal is for TurboQuant-Pro to become the system that can answer:

> **Which quantizer is appropriate for this tensor, this consumer, this workload, and this budget — and what evidence supports that decision?**

This would make third-party methods such as NF4, FP8, NVFP4, GPTQ, AWQ, and future codecs first-class execution backends behind a common consumer-aware decision layer.

---

## Motivation

TurboQuant-Pro already has the ingredients of this architecture.

Its governing principle is consumer-aware compression: compression should be accepted or rejected using the metric consumed downstream — recall, perplexity, routing stability, rank fidelity, or another consumer-relevant measure — rather than reconstruction cosine alone.

The current library already includes several pieces of the control loop:

* quantizer plugin and conformance infrastructure;
* operator tracing and operator sensitivity;
* consumer-aware probes;
* rank certificates;
* runtime fallback policy;
* replayable claims;
* quality monitoring;
* compressed vector indexes;
* KV-cache compression;
* model-weight compression;
* vLLM and database integrations.

The plugin protocol already provides an important foundation: external quantizers can register through the same interface as in-tree implementations and can be tested against a shared conformance contract.

This suggests a larger role for TurboQuant-Pro.

Instead of competing only at the codec level:

```text
TurboQuant-Pro vs NF4 vs FP8 vs GPTQ vs AWQ
```

the architecture should become:

```text
                    TurboQuant-Pro

              identify the consumer
                       |
              identify read geometry
                       |
             determine error sensitivity
                       |
              allocate bit / work budget
                       |
               choose codec + policy
              /        |        |       \
           TQP       NF4      FP8      AWQ
             \         |        |       /
                  certify result
                       |
                     deploy
                       |
                    monitor
                       |
               re-plan / fall back
```

In this architecture, another quantizer outperforming a native TQP codec is not a failure.

If TurboQuant-Pro correctly selects that quantizer for the workload, **TurboQuant-Pro has succeeded**.

---

## Core idea

A compression system should optimize the representation for the computation that will consume it.

Observation Theory gives a formal expression of this idea. A consumer induces a read operator \(P_C\), and the relevant distortion is consumer-relative:

$$
d_C = \operatorname{tr}(P_C \Sigma_\delta)
$$

rather than reconstruction distortion alone:

$$
d_{\text{recon}} = \operatorname{tr}(\Sigma_\delta).
$$

Ordinary reconstruction is therefore the special case \(P_C = I\).

The engineering implication is important:

> **Compression should preserve the distinctions the consumer can use, not necessarily the original representation uniformly.**

TurboQuant-Pro already demonstrates why this matters. For KV keys, apparently excellent reconstruction fidelity can coexist with catastrophic downstream behavior, while a representation chosen for the actual attention consumer preserves useful behavior.

The proposed control plane should generalize that principle across quantizers, operators, frameworks, and workloads.

---

## Proposed capability

Introduce a first-class planning object, tentatively:

```python
CompressionPlanner
```

or:

```python
QuantizationControlPlane
```

It should accept:

```python
plan = planner.plan(
    artifact=tensor_or_model,
    consumer=consumer_spec,
    budget=budget,
    candidates=["tqp", "nf4", "fp8", "nvfp4", "gptq", "awq"],
    objective=objective,
)
```

and return a structured plan containing:

```python
CompressionPlan(
    target=...,
    consumer=...,
    operator_regime=...,
    candidate_results=...,
    selected_codec=...,
    selected_parameters=...,
    expected_cost=...,
    expected_quality=...,
    certificate_requirement=...,
    fallback_policy=...,
    evidence=...,
)
```

The important requirement is that the plan explains **why** the codec was selected.

---

### 1. Identify the consumer

Determine what computation actually consumes the representation.

Examples:

* attention score;
* residual stream;
* MoE routing gate;
* SSM state update;
* vector retrieval;
* reranker;
* classifier;
* user-declared custom consumer.

Existing `operator_trace`, `operator_sensitivity`, and read-operator infrastructure should feed this stage.

---

### 2. Determine the relevant distortion

Prefer the strongest available signal, in approximately this order:

1. direct downstream metric;
2. known analytic read operator;
3. measured / recovered read operator;
4. operator-specific proxy validated against the consumer;
5. reconstruction metric only when it is the appropriate consumer metric.

The planner should never silently substitute reconstruction fidelity for a declared downstream objective.

That rule is already central to the current TQP design.

---

### 3. Enumerate candidate codecs

Candidate codecs should come from the plugin registry rather than from hard-coded TQP-only implementations.

Examples could include:

```text
tqp_per_channel
tqp_polar
bnb_nf4
bnb_int8
fp8_kv
nvfp4_kv
gptq
awq
future external codecs
```

Each candidate should expose capabilities such as:

```text
supported targets
bit widths
native dtype
affine decode support
packed representation
hardware requirements
calibration requirements
fused-kernel compatibility
```

---

### 4. Measure candidates against the consumer

For each candidate, evaluate the quantity that actually matters.

Examples:

```text
retrieval         -> recall@k / rank certificate
attention keys    -> attention logits / perplexity / task quality
MoE routing       -> expert-set flip rate / routing margin
SSM decay         -> downstream state error / perplexity
classification    -> decision agreement
custom operator   -> declared read metric
```

A codec with better MSE but worse consumer behavior should lose.

A codec with worse reconstruction but better consumer behavior should be allowed to win.

---

### 5. Optimize resource allocation

The planner should support more than codec selection.

It should eventually choose:

```text
codec
bit width
per-layer allocation
per-tensor allocation
per-direction allocation
hot/cold treatment
reranking depth
exact-preservation regions
outlier treatment
hardware placement
```

The objective can be expressed as a constrained optimization:

$$
\min \text{cost}
$$

subject to:

$$
Q_C \ge Q_{\min}
$$

or conversely:

$$
\max Q_C
$$

subject to:

$$
R \le R_{\max}.
$$

Where appropriate, the read operator should drive nonuniform bit allocation.

---

## Certification

A selected plan should optionally produce a machine-readable certificate.

The certificate should record:

```text
consumer identity
consumer metric
artifact identity
codec identity + version
codec parameters
plugin identity
hardware/software environment
calibration population
quality measurements
quality floors
certificate or confidence result
known limitations
fallback action
```

This extends the existing TQP certificate philosophy rather than inventing a parallel system.

The existing library already supports provenance-bearing certificates, independent verification, replay, and claim tracking.

---

## False-clear protection

The control plane should explicitly measure **false clears**:

> cases in which a cheap or conventional quality metric accepts a compression that the true consumer rejects.

This should become a standard diagnostic attached to candidate evaluation.

TurboQuant-Pro already contains the beginnings of this capability through `false_clear`, including the distinction between harmful false clears and conservative misses.

A strong planner should be able to report:

```text
cosine passed
consumer failed
false_clear = true
```

rather than merely returning a scalar reconstruction score.

---

## Runtime policy

Planning should not end at deployment.

A plan should carry a runtime policy that can respond to:

```text
distribution drift
operator drift
rank-certificate vacuity
consumer changes
context-length changes
routing-margin collapse
quality-monitor alarms
hardware changes
codec incompatibility
```

Possible runtime actions:

```text
proceed
recalibrate
increase bits
switch codec
rerank more
require exact rerank
keep sensitive tensor fp16
disable compression
```

This should integrate with the existing `TQPRuntimePolicy` rather than replace it.

---

## Multi-consumer support

A major extension should support representations read by more than one consumer.

For example, the same stored embedding may feed:

```text
ANN retrieval
reranker
classifier
RAG pipeline
analytics workload
```

The planner should therefore support:

```python
consumers=[retrieval, reranker, classifier]
```

and identify whether one representation can satisfy all consumers or whether separate representations/refinements are preferable.

Longer term, this is where the two-observer and multi-consumer results from Observation Theory may become directly useful.

---

## CLI concept

A natural top-level workflow could be:

```bash
tqp trace model-or-artifact
tqp plan ...
tqp compare ...
tqp certify ...
tqp deploy ...
tqp monitor ...
```

Example:

```bash
tqp plan model \
  --model Qwen2.5-7B \
  --target kv-cache \
  --consumer attention \
  --budget "4 bit" \
  --candidates tqp,nf4,fp8,nvfp4 \
  --objective perplexity
```

Possible output:

```text
Target: KV keys
Consumer: SOFTMAX_SCORE

Candidate        Bits    Consumer metric     Verdict
-----------------------------------------------------
symmetric NF4      4     FAIL                reject
TQP asym-NF4       4     PASS                candidate
FP8                 8     PASS                candidate
NVFP4               4     PASS                candidate

Selected:
  TQP asym-NF4

Reason:
  Best consumer quality satisfying 4-bit budget.

Rejected:
  symmetric NF4 — reconstruction metric acceptable,
                  downstream consumer failed.

Certificate:
  available

Runtime fallback:
  per-channel_or_fp16
```

---

## Plugin ecosystem consequence

This feature should intentionally make TurboQuant-Pro attractive to **other quantizer authors**.

An external project should be able to implement:

```python
class MyQuantizer:
    ...
```

register it through the plugin API, and immediately gain access to:

```text
consumer-aware evaluation
comparison harnesses
certification
replay
runtime monitoring
hardware capability reporting
planner participation
```

That changes TQP from a competing implementation into an ecosystem.

The strategic target should be:

> **If someone invents a great new quantizer, the fastest way to demonstrate where it belongs should be to plug it into TurboQuant-Pro.**

---

## Acceptance criteria

A first production-quality version should satisfy the following:

1. At least three materially different codec families can participate in one planning run.
2. At least two codecs are supplied through the external plugin mechanism.
3. Codec selection is based on a declared consumer metric rather than reconstruction fidelity.
4. The planner can select a third-party codec over a native TQP codec.
5. The reason for the selection is machine-readable.
6. A rejected codec can explicitly report a false-clear condition.
7. The selected plan can produce a certificate or evidence bundle.
8. The plan includes a declared runtime fallback.
9. Re-running the same plan against the same identified artifact and environment is reproducible.
10. Unknown or unsupported consumer regimes fail conservatively rather than inventing a recommendation.

---

## Non-goals

This feature should **not** require TurboQuant-Pro to:

* implement every quantization algorithm itself;
* claim one codec wins universally;
* replace TensorRT-LLM, vLLM, bitsandbytes, GPTQ, AWQ, or other runtimes;
* infer safety from reconstruction metrics;
* silently recommend a codec in an unmeasured regime.

The control plane should prefer:

```text
ABSTAIN
```

to unsupported certainty.

---

## Strategic objective

The long-term goal is larger than building another quantization package.

TurboQuant-Pro should become:

> **the system that determines which representation is sufficient for a specified consumer, allocates the available compression budget accordingly, selects the appropriate implementation, certifies the result, and monitors whether the original assumptions remain valid.**

In that model, the primitive codec is only one component.

The product is the **decision**.

The moat is the combination of:

```text
consumer modeling
measurement
codec interoperability
planning
certification
runtime adaptation
evidence
```

A successful implementation would make TurboQuant-Pro less like a single quantization algorithm and more like an **optimizer and control plane for lossy representation**.

That is the level at which TurboQuant-Pro should aim to become the best compression library.

**Comment by ahb-sjsu** [comment]
## P0 is implemented and merged to master in #170

The design this issue describes was already written up in `docs/DESIGN_planner.md`, and every ingredient it lists was already in the tree. What was missing was the contract that joins them. That is now built and on master as of `de6729a`.

This does **not** close the issue. P0 is one milestone of four, and the list of what is still missing is at the bottom.

### What landed

**`turboquant_pro/planner.py`**  `CompressionPlanner`, aliased `QuantizationControlPlane` as this issue names it.

A `WorkloadSpec` declares the artifact, consumer, budget and quality floor. Preflight measures the data preconditions that change what is legal. Candidates are enumerated **from the plugin registry**, so a third-party codec competes on the same terms as an in-tree one. Priors on stored bytes rule out what cannot fit. Survivors are measured on the consumer's own metric over a calibration split with successive halving. The frontier is kept under uncertainty rather than collapsed by a mean. The winner is verified once on a held-out split that played no part in the search.

**`turboquant_pro/consumers.py`**  the third extension point, completing the pair of `plugins` (the codec) and `read_operators` (the operator it is judged against): the per-item measurement that stands in for the consumer. Ships:

| consumer | per item | direction |
|---|---|---|
| `topk_inner_product` / `topk_cosine` / `topk_l2` | recall@k of a search over the compressed corpus, per query | higher better |
| `attention_softmax` | logit rank agreement or softmax KL over the keys, per query | mode dependent |
| `attention_topk` | fraction of the exact top-k attended keys still attended | higher better |
| `read_operator` | `delta_i^T P_C delta_i`, whose mean is `tr(P_C Sigma_delta)` | lower better |
| `declared` | whatever the caller supplies | declared |

It has its own entry-point group, `turboquant_pro.consumers`, so a consumer can arrive from out of tree exactly like a codec.

Also: `plugins.capabilities()`, a plan-record JSON schema, and `tqp plan run | explain | replay | consumers`.

### Against the acceptance criteria

1. **Three codec families in one run**  yes. Candidates come from the registry, not a list.
2. **Two codecs through the plugin mechanism**  yes, and the tests register theirs the same way an out-of-tree package does.
3. **Selection on a declared consumer metric**  yes. `selection.decided_on` names the metric and `selection.not_decided_on` says "reconstruction cosine (diagnostic only)".
4. **A third-party codec can beat a native one**  yes, with a test that makes it happen on KV keys against `per_channel` and asserts `selection.third_party`.
5. **Machine-readable reason**  `selection.rule`, `selection.reason`, `selection.beat` with the margin, and per candidate a `verdict` and the `left_at` stage.
6. **A rejected codec reports a false clear**  yes, and the diagnostic is attached to **every** candidate evaluation rather than only the winner's. It is scored on the consumer's own items: a cosine over corpus rows compared against a recall over queries is two populations, so when they cannot be aligned the plan says the diagnostic is unavailable rather than computing a misleading one.
7. **Certificate or evidence bundle**  evidence entries tagged `statistical`, `certificate`, `measured_cost`, `diagnostic`, `comparison`, each naming its population and method.
8. **A declared runtime fallback**  actions and thresholds come from `runtime_policy.TQPRuntimePolicy`, not a parallel vocabulary, with the triggers that would make the plan stale.
9. **Reproducible**  `replay_plan` re-runs the verification and reports agreement. A replay against different bytes reports the mismatch rather than claiming reproduction.
10. **Unknown regimes fail conservatively**  an unregistered consumer, a consumer defined for another target, or a floor nothing reaches gives a plan whose `selected_codec` is the literal `ABSTAIN`, with the reason and the fallback intact. `tqp plan run` exits 1.

One addition not in the list, because the campaign kept paying for its absence: **quality is accepted on the conservative end of a bootstrap interval, never the mean.** A mean that clears a floor by less than its own noise has not cleared it.

### What it looks like

```
codec                  bits   bytes/vec    consumer (bound)  verdict
ext_uniform/2             2       104.0              0.6570  reject
ext_uniform/8             8       104.0              0.9880  selected
ext_flatterer/8           8       104.0              0.8060  reject

selected: ext_uniform {'bits': 8}
  rule:   among frontier candidates meeting every budget, maximise the
          consumer metric's conservative bound, then minimise stored bytes
  reason: recall@10/inner_product bound 0.9880 on held-out items at 95%;
          floor 0.85; 104.0 stored bytes/vector
  ! ext_flatterer{'bits': 8}: the cheap metric cleared it and the consumer
    did not (P(consumer fails | cosine cleared) = 0.62)
certificate: certified  distribution-free rank floor holds for this corpus
runtime fallback: rerank_more
```

### Three bugs the tests found

Worth recording, because each was a wrong answer waiting to happen.

- **Successive halving was cutting candidates that still cleared the floor.** The floor is the bar; quality above it is a tiebreak. Under `min_cost` this meant the cheapest acceptable codec was eliminated during the search, before cost had ever been consulted. Fixed: a candidate that clears the floor is never cut early.
- **A consumer config key called `name` collided with the registry's own parameter.** `declared` has one, for the metric's label. The registry name is now positional-only in both `create_consumer` and `plugins.create`.
- **Preflight's spectrum was computed over non-finite rows**, returning NaN and a numpy warning on exactly the data that most needs a preflight. Non-finite rows are now excluded from the spectrum and still counted in the report.

### What is not done

Named here rather than implied:

- No measured latency or throughput evidence. `measured_cost` currently covers size and not time.
- A candidate is a codec, not yet a pipeline. Transforms and search operators are not candidate stages.
- No faiss or rabitqlib adapters.
- No runtime loop (canaries, drift triggers, re-plan scheduling). The fallback is declared in the record but nothing yet acts on it.
- **The preregistered regret exit test against the RaBitQ campaign grid has not run.** Until it does this is a working control plane, not a validated one. A planner that recommends is making a claim, and this one has not yet been scored as one. Tracked in #172.

### Verification

Public CI on #170, all six lanes green:

| lane | result |
|---|---|
| test (3.10) | pass, 1m35s |
| test (3.11) | pass, 1m53s |
| test (3.12) | pass, 2m3s  **1831 passed, 111 skipped** |
| test-py39 | pass, 1m13s |
| torch-cpu | pass, 59s |
| adc-kernel | pass, 28s |

The 3.12 lane went from 1678 passed on master before this to 1831, which is the 153 new tests running. `ruff` and `black` are clean; `ty` is advisory in this workflow and the two new modules add **zero** diagnostics to its count. Every one of these lanes was also run locally before the PR was opened, including the Python 3.9 lane in a container, so CI was confirming rather than discovering.

The new tests are `tests/test_planner.py` (the acceptance criteria above, end to end), `tests/test_consumers.py` (the consumer contract, registry semantics, entry-point resilience and the error paths that must raise rather than substitute a metric nobody asked for) and `tests/test_planner_internals.py` (byte accounting over nested containers, dataclasses, slots, shared buffers, cycles and sparse overlays; preflight; the bootstrap bound in both directions; dominance; the shape-hint convention; and every path where something fails and has to be recorded rather than dropped).

### One thing found on the way that is not part of this change

Building the AVX2 ADC kernel with `-march=native` on a Xeon E5-2690 v3 and then running the whole suite, four tests fail that pass without it:

```
tests/test_index.py::test_mmap_search_matches_in_ram
tests/test_index.py::test_v3_roundtrip_identical_ram_and_mmap
tests/test_index.py::test_cosine_unchanged_by_the_l2_refactor
tests/test_ivf.py::test_probe_all_equals_bruteforce   (recall 0.978 vs exact)
```

They fail identically on the parent commit, so this is not a regression from the planner. It is invisible to CI because the `adc-kernel` job builds the kernel and then runs only `tests/test_adc_kernel.py`, which passes; the matrix lane that runs `tests/` never has the kernel present, so those files skip. Filed as #171.
